core/src/main/java/org/elasticsearch/plugins/Plugin.java

```diff
@@ -221,8 +221,9 @@ public void onIndexModule(IndexModule indexModule) {}
      * interface. In this case we assume the node is running in production and all bootstrap checks must pass. This allows plugins
      * to provide a better out of the box experience by pre-configuring otherwise (in production) mandatory settings or to enforce certain
      * configurations like OS settings or 3rd party resources.
+     * @param metaData the on-disk metadata for this nodes. This allows bootstrap-checks based on the locally recovered cluster metadata
      */
-    public List<BootstrapCheck> getBootstrapChecks() { return Collections.emptyList(); }
+    public List<BootstrapCheck> getBootstrapChecks(MetaData metaData) { return Collections.emptyList(); }
```

Comment:
I think that I would rather see this pushed through at check execution time (it's not fair this is only exposed to plugin checks). We also have some checks that rely on settings and we do this at construction time instead of check time so maybe the change here to add a `BootstrapCheckContext` that carries settings and this metadata, check pushes this through to each check and they can make decisions  on that basis?

Reply:
so my biggest concern here is the way bootstrap checks work today. we have a method that returns a boolean and a method that returns a message. This nature makes it very hard to give a good message bases on the state while if I have all information at construction time its simple to prepare the message. The settings is another issue since it might access settings that are coming from the keystore that should be accessed ahead of time (before we execute them) I think there are many issues here that I think we should address. The internal bootstrap checks can also get this information if needed we can change the protected node method?

Reply:
The return type of `BootstrapCheck#check` can easily be changed away from being a boolean?

Reply:
yeah I wonder if that is a bit too much of a refactoring give this goes into 6.0?

Reply:
I hear that concern and while it will be a lot of lines of code changed, I think it can be done quickly and is not a de-stabilizing change.